### ISSUE TITLE
Code Segment is not restored after far call on x86

### ISSUE THREAD

**Issue by GregoryMorse** [NationalSecurityAgency/ghidra#765]
**Describe the bug**
The x86 (important for 16-bit programs mainly) CS register is not recovered in the context properly by sleigh (C version) after a far call is executed which has a different segment.  The loader currently attempts to place imports in the same segment so this bug is less likely to turn up.  Upon continuing disassembly scanning after a far call, the CS register should be implicitly restored to the current instruction pointer/program counter (IP) execution address which requires that CS to even do the scanning.  But the scanning proceeds without even needing the context CS:IP values so until an instruction requests CS explicitly its not an issue.  Jump tables specifically will cause this problem.

**To Reproduce**
Decompile code with the following sequence:
...
CALLF <import> ; 0x9a
<code...>
JMP CS:[offset+bx]
offset: <jump table...>
<code for jump table offsets...>

**Expected behavior**
sleigh should return the CS register after a far call to the value prior to the call which is the current CS:IP pair of the code being scanned.  The Program Counter in this case is really the pair CS:IP and not merely IP which is part of the problem.  Due to the thoroughness of the architecture I have not yet pinpointed exactly how to fix this problem.

**Environment (please complete the following information):**
 - OS: Windows 10
 - Java Version: 11.0.3
 - Ghidra Version: 9.0.4

**Comment by emteere** [comment]
Normally the segments are mapped on top of a continuous set of bytes, especially if they overlap and the bytes are the same at the overlap.  Set correctly, the normal segment address calculations will address the correct byte.  I believe the 16-bit NE loader works this way.

I haven't looked at 16-bit in a while, but I believe the CS segment is set for each loaded segment.
For your segments you have created, you can set the CS register across each segment to the correct value.  The decompiler will take the CS value set at the start of the function and should keep the same value across calls.  The CS register is listed as "unaffected" by calls in the 16-bit .cspec calling conventions.

The CS shouldn't actually be set in the far call, and if it is, it should be restored in the pcode.

**Comment by GregoryMorse** [comment]
Yes your understanding is correct, there is overlap, and the CS is marked as unaffected in the cspec.  The unaffected marking might be a first place to track this down.

The jump table using the CS register to access data is the only reason this register ends up being used as executables often put the jump table right beneath the jump instruction.

AFAIK, the processor spec says the CS is set in the far call.  I don't know how it is restored - the unaffected marking should be respected.  Or it should ask for a "getCallFixup" - some Pcode which restore it.  But for sure the decompiler is not asking for the call fixup.  Yet since it is already marked as unaffected may it should not have to.

My understanding then is the bug lies in that:
1) the decompiler should respect the unaffected marking of the CS register and not have a stale context.
2) the decompiler should request via getCallFixup some Pcode which can send in the correct unaffected register values.

Would like to fix and submit a pull request for this but due to the sophistication of Ghidra's architecture, I am simply not yet sure where the root of this issue is.

**Comment by GregoryMorse** [comment]
If can find one, will add a real world example of this bug with a binary sample and xml debug info.  Right now its contrived as custom loader and sleigh translator are used which does not resolve fixups deliberately - though theoretically this should be no issue.  The fixup resolution in Ghidra is merely masking the bug as imports should be able to use any placeholder code segment instruction pointer pair one wishes to use.

**Comment by GregoryMorse** [comment]
This still seems to be a bug with unaffected registers.  I have learned more about the Ghidra architecture:

I see there is a fixup of CALLMECHANISM_TYPE to deal with generating Pcode across caller-callee boundaries.  However, x86-16 has no fixups for this case nor even does the 32-bit module have any except for some special calls.  But the CS register is hardly ever used directly except for jump tables and only in a case where it previously made a far call to a different code segment.  I can interact with the decompiler in a custom manner to highlight this but Ghidra always remaps imports to the code segment.  I have not found this in a large app - again the specific sequence needed may be relatively rare.

Furthermore as was stated, when a register is marked unaffected, no call fixup is needed.  Basically the unaffected flag is for whatever reason being ignored so when a jump table addressed by CS:[REG+OFFSET] is hit, it wrongly tracks the CS back to the last far call.

Since the far call changes CS correctly, the jump table reads the CS incorrectly but still tries to use CS for sure, and it is marked unaffected, there is one conclusion: the decompiler has a bug with unaffected registers across call boundaries in some scenarios - maybe specifically the CS register since its part of the execution flow.

**Comment by GregoryMorse** [comment]
I would also mention that as is noted in other issues reported here, something clearly wrong with the CS register analysis is occurring - its basically seeing the stack argument as CS:IP and not the pushed values:

A snippet as:
```
push 0
push 0
call CLIPCURSOR
call RELEASECAPTURE
retn 4
```
Yields:
```
void func(void)
{
  undefined2 unaff_CS;
  
  CLIPCURSOR((tagRECT *)CONCAT22(unaff_CS,0x49));
  RELEASECAPTURE();
  return;
}
```

When clearly the right decompilation is:
```
void func(void)
{
  CLIPCURSOR(nullptr);
  RELEASECAPTURE();
  return;
}
```
One bug fix in the right place probably clears out 3 to 4 outward bugs.

**Comment by GregoryMorse** [comment]
Conclusion:
1) decompiler does not respect CS register as unaffected when wrong calling convention is chosen and far call makes it think it is a stack argument.
2) decompiler does not query external function information for far calls perhaps due to the former - which does not give the Java calling convention analyzer a chance to even specify the model!
3) presumably the model would be returned correctly by calling convention analyzer - but we cannot test this without the prior to bugs fixed.

I noticed x86win.cspec has something like this which might help resolve the issue:
```
  <resolveprototype name="__fastcall/__thiscall/__stdcall">
     <model name="__stdcall"/>      <!-- The default case -->
     <model name="__fastcall"/>
     <model name="__thiscall"/>
  </resolveprototype>
  <eval_current_prototype name="__fastcall/__thiscall/__stdcall"/>
```

The calling convention information in the x86-16.cspec file is pretty weak and hopefully the source of all the troubles.  Perhaps and likely it does not show up on 32 or 64 bit because of mitigations based on the cspec.

Nonetheless the decompiler would be far more resilient if it could handle both 1) and 2).

**Comment by GregoryMorse** [comment]
More technical details to solve this issue found in ia.sinc - it seems to verify the Sleigh code looks correct as it assigns the CS and uses an offset of CS*16+offset which is best approach:
# The far addresses listed here actually specify the CS segment to use
# but we do not model changing the CS segment in protected mode
# so we just use the offset portion of the address
```
ptr1616: reloc is imm16; j16		[ reloc = j16*16 + imm16; ] { CS = j16; export *[ram]:4 reloc; }
ptr1632: j16":"imm32 is imm32; j16	{ CS = j16; export *:4 imm32; }

:CALLF ptr1616      is vexMode=0 & addrsize=0 & opsize=0 & byte=0x9a; ptr1616           { push22(CS); build ptr1616; push22(&:2 inst_next); call ptr1616; }
:CALLF ptr1616      is vexMode=0 & addrsize=1 & opsize=0 & byte=0x9a; ptr1616           { push42(CS); build ptr1616; push42(&:2 inst_next); call ptr1616; }
:CALLF ptr1632      is vexMode=0 & addrsize=0 & opsize=1 & byte=0x9a; ptr1632           { push22(CS); build ptr1632; push24(&:4 inst_next); call ptr1632; }
:CALLF ptr1632      is vexMode=0 & addrsize=1 & opsize=1 & byte=0x9a; ptr1632           { push42(CS); build ptr1632; push44(&:4 inst_next); call ptr1632; }
```
It basically looks like the CS register is not being dealt with properly.  It will only work if the far call is in the same segment as it uses the offset only.  No wonder the rest of the errors occur.

The question is probably how to fix the sleigh code?

**Comment by GregoryMorse** [comment]
@emteere - the unaffected CS in the compiler spec seems to be not yet known by the decompiler when it encounters the code.  The decompiler does not ask for external reference information until it is towards the end of decompilation and for x86-16 bit it never does at all due to a bug with it only doing it for indirect calls.  At that point it would know what is unaffected but it already had scanned the jump table after the far call with a bad CS register.  If the external function was provided as if it was a local function it can help to resolve this issue.  But nonetheless, the decompiler cannot just keep scanning without having determined that special case CS register is restored.  It should not rely on the compiler spec as the CS register is special to execution and continuing decompilation would be technically not possible without CS being set to its only logical value.  The other unaffected registers can have postponed analysis but not this one.

This is a bug inside of the decompiler which needs to have a better concept of segmentation - from the processor spec level.

**Comment by GregoryMorse** [comment]
I have worked around this bug by putting in the cspec file the following in the __stdcall16far prototype:
<pcode inject="uponreturn" dynamic="true"/>

Then I inject Pcode which simply takes the offset, gets the segment, and uses that segments selector (so it does the translation) and emit pcode dynamically:
```CS = <get_selector(get_segment(offset))>;```

And the jump tables are processed properly thereafter.  The concept of the CS register needs better integration in Ghidra.  This work around is hardly ideal

**Comment by Wall-AF** [comment]
@emteere This seems to have been forgotten.  Can this be re-examined to decide if it has been rejected or catered for (in another way - if so, please elaborate)?
